qubeswhonix/__init__.py

```diff
@@ -18,92 +18,185 @@
+            # If VM has no default_dispvm, assume user doesn't want one.
+            default_dispvm = None
+
+        elif template_for_dispvms:
+            # If VM is a Workstation DVM, use itself.
+            default_dispvm = vm
+
+        elif (
+            curr_default_dispvm
+            and curr_default_dispvm.features.check_with_template(
+                "whonix-ws", None
+            )
         ):
-            default_dispvm = template.features['whonix-default-dispvm']
-        elif template is not None:
-            #  example template.name: whonix-ws-14
-            # example default_dispvm: whonix-ws-14-dvm
-            default_dispvm = template.name + '-dvm'
-        else:
-            # assume whonix-workstation-17-dvm is right
-            # HARDCODED.
-            default_dispvm = 'whonix-workstation-17-dvm'
+            # If VM's default_dispvm is Whonix, use it.
+            default_dispvm = curr_default_dispvm
 
-        if default_dispvm in app.domains:
-            vm.default_dispvm = default_dispvm
         else:
-            # expected default dispvm does not exists, log an error and set
-            # default dispvm to None
-            vm.log.error('QubesWhonixExtension: default dispvm\'%s\' does '
-                         'not exists', default_dispvm)
+            # Standalones don't have a template, return thyself.
+            template = getattr(vm, "template", vm)
+            default_dispvm = get_template_dispvm(template)
+
+        set_default_dispvm(vm, default_dispvm)
+
+    @staticmethod
+    def set_gw_dispvm(app, vm):
+        """Set the default DispVM for a Whonix-Gateway qube to None if the
+        current one is not a Whonix-Workstation."""
+        # pylint: disable=unused-argument
+        default_dispvm = getattr(vm, "default_dispvm", None)
+        if not default_dispvm:
+            return
+        if not default_dispvm.features.check_with_template("whonix-ws", None):
             vm.default_dispvm = None
 
-    @qubes.ext.handler('domain-add', system=True)
+    @qubes.ext.handler("domain-add", system=True)
     def on_domain_add(self, app, _event, vm, **_kwargs):
-        '''Handle new AppVM created on whonix-ws/whonix-gw template and
+        """Handle new AppVM created on whonix-ws/whonix-gw template and
         adjust its default settings
-        '''
-        template = getattr(vm, 'template', None)
+        """
+        template = getattr(vm, "template", None)
         if template is None:
             return
 
-        if 'whonix-gw' in template.features:
-            vm.tags.add('anon-gateway')
-            vm.tags.add('sdwdate-gui-server')
+        if "whonix-gw" in template.features:
+            vm.tags.add("anon-gateway")
+            vm.tags.add("sdwdate-gui-server")
+            self.set_gw_dispvm(app, vm)
 
-        if 'whonix-ws' in template.features:
+        if "whonix-ws" in template.features:
             # this is new VM based on whonix-ws, adjust its default settings
```

Comment:
This was a problem with the original code too, but using constructs like `if 'whonix-gw' in template.features` means that the appropriate tags (and also appropriate NetVMs/DispVMs) won't be set on StandaloneVMs, since they have no template. I think we can port this to using `vm.features.check_with_template()`, which should work. (As a side-effect, if an AppVM is marked as `whonix-gw`, that will also be honored. I don't think that's undesirable behavior though.)